scripts/prettier/prettier-helpers.js

```diff
@@ -0,0 +1,109 @@
+let projectsWithPrettierConfig;
+
+function init() {
+  if (projectsWithPrettierConfig) {
+    return;
+  }
+
+  projectsWithPrettierConfig = [];
+  const rushJson = readConfig('rush.json');
+  if (rushJson) {
+    // Check the root of each project for a custom prettier config, and save the project paths that have one
+    for (const project of rushJson.projects) {
+      const packagePath = path.resolve(repoRoot, project.projectFolder);
+      if (fs.existsSync(path.join(packagePath, prettierConfig))) {
+        projectsWithPrettierConfig.push(packagePath);
+      }
+    }
+  }
+}
+
+/**
+ * Run prettier for a given set of files with the given config.
+ *
+ * @param {string[]} files List of files for which to run prettier
+ * @param {string} configPath Path to relevant prettier.config.js.
+ * @param {boolean} [runAsync] Whether to run the command synchronously or asynchronously
+ * @returns A promise if run asynchronously, or nothing if run synchronously
+ */
+function runPrettier(files, configPath, runAsync) {
+  const cmd = `node ${prettierBin} --config ${configPath} --ignore-path "${prettierIgnorePath}" --write ${files.join(' ')}`;
+  if (runAsync) {
+    return exec(cmd, undefined, undefined, process);
+  } else {
+    execSync(cmd);
+  }
+}
+
+/**
+ * Runs prettier on all ts/tsx/json/js files in a project.
+ *
+ * @param {string} projectPath Path to the project root for which to run prettier
+ */
+function runPrettierForProject(projectPath) {
+  init();
+
+  const sourcePath = path.join(projectPath, '**', '*.{ts,tsx,json,js}');
+  const configPath = projectsWithPrettierConfig[projectPath] || prettierRulesConfig;
+  runPrettier([sourcePath], configPath);
+}
+
+/**
+ * Runs prettier on the given list of files.
+ *
+ * @param {string[]} files Staged files passed in by lint-staged
+ * @param {boolean} [runAsync] Whether to run the command synchronously or asynchronously
+ * @returns A promise if run asynchronously, or nothing if run synchronously
+ */
+function runPrettierMultiProject(files, runAsync) {
+  if (files.length === 0) {
+    return runAsync ? Promise.resolve() : undefined;
+  }
+
+  init();
+
+  // Buid a mapping from config file name to files for which that config applies
+  const configMap = {};
+  for (const file of files) {
+    // Default to the repo-wide config
+    let configPath = prettierRulesConfig;
+    const absPath = path.resolve(repoRoot, file);
+    for (const projectPath of projectsWithPrettierConfig) {
+      // Check if this file is inside any of the projects with a custom config
+      if (absPath.startsWith(projectPath)) {
+        configPath = path.join(projectPath, prettierConfig);
+        break;
+      }
+    }
+    if (!configMap[configPath]) {
+      configMap[configPath] = [];
+    }
+    configMap[configPath].push(file);
+  }
+
+  const configPaths = Object.keys(configMap);
+  console.log('CONFIG PATHS: ' + configPaths);
+  console.log(JSON.stringify(configMap));
+  let promise = Promise.resolve();
```

Comment:
This will actually run every prettier command on all the files, but it will only return the last promise. Sometimes the promises don't resolve at right order because of async. You probably should do one of two things:

1. return a `Promise.all(configPaths.map(configPath => ...))` - this will run all prettiers at the same time

or

2. Use something like the `p-limit` for something like this: https://www.npmjs.com/package/p-limit to limit the rate at which you are running your promises in parallel

Reply:
What I intended to do was run them all in serial:
```
let promise = Promise.resolve();
for (const thing of things) {
  promise = promise.then(() => doStuff(thing));
}
```
Except the `.then` part got lost in refactoring... (also might not work quite right since the sync version of the method returns undefined and the async version returns a promise)

Reply:
Updated with the serial pattern I meant to use...